### PR TITLE
dm/video-image-training

### PR DESCRIPTION
**Pull request by danielmend**
- allow us to intermix image embeddings with video frame embeddings in training for higher data diversity

### CODE REVIEW

**Comment by danielmend** [comment]
todo:

- [ ] batch size arg for image training
- [ ] figure out how to get rid of EmbeddingReader loading bar
- [ ] ensure image/text pairs are read correctly (i.e. the pairs we say match actually match)
- [ ] figure out data shuffling // what subset of laion5b should we be training on?
- [ ] can i fit both image and text embeddings into one EmbeddingReader?

File: src/training/train.py [hunk 1]
```diff
@@ -20,28 +22,66 @@ def train_one_epoch(model_video, data, epoch, optimizer, scheduler, args, tb_wri
         world_size=1,
         use_horovod=False,
     )
+
     dataloader = data["train"].dataloader
+    if args.image_data:
```
Comment (iejMac) [inline src/training/train.py:27]: for now I guess this is fine but ideally we want to move this into data.py (I know it's bad) and only create it once in main.py then pass into this function

File: src/training/train.py [hunk 2]
```diff
@@ -20,28 +22,66 @@ def train_one_epoch(model_video, data, epoch, optimizer, scheduler, args, tb_wri
...
+            vid_emb = torch.zeros_like(embeddings)
+            
+            vid_emb[:, 0, :] = img_embeddings
+            vid_emb = vid_emb.to(device, non_blocking=True)
+
+            text_embeddings, _ = next(text_iter)
+            text_embeddings = torch.tensor(text_embeddings)
+            text_embeddings = text_embeddings.to(device, non_blocking=True)
+
+            video_embeddings = model_video.encode_video(embeddings, prenorm=True, postnorm=True)
```
Comment (iejMac) [inline src/training/train.py:75]: I think instead of embeddings this should be vid_emb

File: src/training/data.py [hunk 3]
```diff
@@ -361,5 +363,34 @@ def get_data(args, preprocess_fns, epoch=0):
...
+    if args.image_data:
+        embeddings_images = EmbeddingReader(
+            embeddings_folder=f'{args.image_data}/img_emb/',
+            file_format='npy'
+        )
+        embeddings_txt = EmbeddingReader(
+            embeddings_folder=f'{args.image_data}/text_emb/',
+            file_format = 'npy'
+        )
+        worker_start_indices = torch.linspace(0, 1e9, args.world_size, dtype=torch.long)
```
Comment (iejMac) [inline src/training/data.py:375]: instead of 1e9 shouldn't this be something like args.train_num_samples or something like that?

File: src/training/data.py [hunk 4]
```diff
@@ -361,5 +363,34 @@ def get_data(args, preprocess_fns, epoch=0):
...
+        worker_start = worker_start_indices[args.rank].item()
```
Comment (iejMac) [inline src/training/data.py:376]: make sure args.rank is global rank

File: src/training/data.py [hunk 5]
```diff
@@ -361,5 +363,34 @@ def get_data(args, preprocess_fns, epoch=0):
...
+        img_iter = iter(
+            embeddings_images(
+                batch_size=args.image_batch_size,
+                start=worker_start,
+                end=worker_start + data["train"].dataloader.num_batches*args.image_batch_size,
```
Comment (iejMac) [inline src/training/data.py:381]: I think you also need to divide by world_size

File: src/training/main.py [hunk 6]
```diff
@@ -125,8 +124,8 @@ def main():
     elif args.report_to == "wandb" and is_master(args):
         logging.debug("Starting wandb.")
         wandb.init(
-            project="video-clip",
-            entity="iejmac", #TODO: do you need this?
+            project="laion-video-clip",
+            entity="danielmend", #TODO: do you need this?
```
Comment (iejMac) [inline src/training/main.py:128]: maybe just make these blank or something

File: src/training/train.py [hunk 7]
```diff
@@ -1,58 +1,169 @@
...
 
-def train_one_epoch(model_video, data, epoch, optimizer, scheduler, args, tb_writer=None):
+from torch.cuda.amp import autocast
+def train_one_epoch(model_video, data, epoch, optimizer, scaler, scheduler, args, tb_writer=None):
     device = torch.device(args.device)
     model_video.train()
     loss_func = ClipLoss(
         local_loss=False,
-        gather_with_grad=False,
+        gather_with_grad=True,
```
Comment (iejMac) [inline src/training/train.py:19]: should get args.gather_with_grad or something
